internal/containerd/restart_unix.go

```diff
@@ -22,35 +22,146 @@ package containerd
+		}
+	}
+
+	return nil
+}
+
+type K0sRestarter struct{}
+
+func (c K0sRestarter) Restart() error {
+	// First, collect systemd units to determine which mode k0s is running in, eg
+	// k0sworker or k0scontroller
+	units, err := nsenterCmd("systemctl", "list-units").CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("unable to list systemd units: %w", err)
+	}
+	service := regexp.MustCompile("k0sworker|k0scontroller").FindString(string(units))
+
+	out, err := nsenterCmd("systemctl", "restart", service).CombinedOutput()
+	slog.Debug(string(out))
+	if err != nil {
+		return fmt.Errorf("unable to restart %s: %w", service, err)
+	}
+
+	return nil
+}
+
+type K3sRestarter struct{}
+
+func (c K3sRestarter) Restart() error {
+	// This restarter will be used both for stock K3s distros
+	// using systemd as well as K3d, which does not.
+	if UsesSystemd() {
+		out, err := nsenterCmd("systemctl", "restart", "k3s").CombinedOutput()
+		slog.Debug(string(out))
+		if err != nil {
+			return fmt.Errorf("unable to restart k3s: %w", err)
+		}
+	} else {
+		// TODO: this approach still leads to the behavior mentioned in https://github.com/spinframework/runtime-class-manager/issues/140:
+		// The first pod's provisioner container exits with code 255, leading to pod status Unknown,
+		// followed by the subsequent pod's provisioner container no-op-ing and finishing with status Completed.
+		pid, err := getPid("k3s")
+		if err != nil {
+			return err
+		}
+		slog.Debug("found k3s process", "pid", pid)
+
+		err = syscall.Kill(pid, syscall.SIGHUP)
+		if err != nil {
+			return fmt.Errorf("failed to send SIGHUP to k3s: %w", err)
+		}
+	}
+
+	return nil
+}
+
+type MicroK8sRestarter struct{}
+
+func (c MicroK8sRestarter) Restart() error {
+	out, err := nsenterCmd("systemctl", "restart", "snap.microk8s.daemon-containerd").CombinedOutput()
+	slog.Debug(string(out))
 	if err != nil {
-		return err
+		return fmt.Errorf("unable to restart snap.microk8s.daemon-containerd: %w", err)
 	}
-	slog.Debug("found containerd process", "pid", pid)
 
-	err = syscall.Kill(pid, syscall.SIGHUP)
+	return nil
+}
 
+type RKE2Restarter struct{}
+
+func (c RKE2Restarter) Restart() error {
+	// First, collect systemd units to determine which mode rke2 is running in, eg
+	// rke2-agent or rke2-server
+	units, err := nsenterCmd("systemctl", "list-units").CombinedOutput()
 	if err != nil {
-		return fmt.Errorf("failed to send SIGHUP to containerd: %w", err)
+		return fmt.Errorf("unable to list systemd units: %w", err)
 	}
+	service := regexp.MustCompile("rke2-agent|rke2-server").FindString(string(units))
+
+	out, err := nsenterCmd("systemctl", "restart", service).CombinedOutput()
+	slog.Debug(string(out))
+	if err != nil {
+		return fmt.Errorf("unable to restart %s: %w", service, err)
+	}
+
 	return nil
 }
 
-func getPid() (int, error) {
+// TODO: lifted and amended from https://github.com/spinframework/runtime-class-manager/pull/387
```

Comment:
You can remove this line :). I am now going to lift your revised version but I'll likely merge after you

Reply:
Thanks for the inspo on this one -- I actually refactored a bit in the last commit.  There were a few uses for listing systemd units already and I realized I could use the same functionality to test systemd being used on the host for eg k3d as well.  WDYT?

Reply:
Nice. I like how we are reusing the same command.